drivers/overwrite_queue.go

```diff
@@ -0,0 +1,119 @@
+package drivers
+
+import (
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/livepeer/go-livepeer/common"
+)
+
+const (
+	timeoutMultiplier = 1.5
+)
+
+type (
+	OvewriteQueue struct {
+		desc           string
+		name           string
+		session        OSSession
+		maxRetries     int
+		initialTimeout time.Duration
+		maxTimeout     time.Duration
+		queue          chan []byte
+		quit           chan struct{}
+	}
+)
+
+func NewOverwriteQueue(session OSSession, name, desc string, maxRetries int, initialTimeout, maxTimeout time.Duration) *OvewriteQueue {
+	oq := &OvewriteQueue{
+		desc:           desc,
+		name:           name,
+		maxRetries:     maxRetries,
+		session:        session,
+		initialTimeout: initialTimeout,
+		maxTimeout:     maxTimeout,
+		queue:          make(chan []byte, 32),
+		quit:           make(chan struct{}),
+	}
+	if maxRetries < 1 {
+		panic("maxRetries should be greater than zero")
+	}
+	go oq.workerLoop()
+	return oq
+}
+
+// Save queues data to be saved
+func (oq *OvewriteQueue) Save(data []byte) {
+	oq.queue <- data
+}
+
+// StopAfter stops reading loop after some time
+func (oq *OvewriteQueue) StopAfter(pause time.Duration) {
+	go func(p time.Duration) {
+		time.Sleep(p)
+		close(oq.quit)
+	}(pause)
+}
+
+// waitForQueueToClear used in tests
+func (oq *OvewriteQueue) waitForQueueToClear(timeout time.Duration) {
+	// wait for work to be queued
+	time.Sleep(5 * time.Millisecond)
+	start := time.Now()
+	for {
+		if len(oq.queue) == 0 {
+			return
+		}
+		if time.Since(start) > timeout {
+			return
+		}
+		time.Sleep(20 * time.Millisecond)
+	}
+}
+
+func (oq *OvewriteQueue) workerLoop() {
+	var err error
+	var took time.Duration
+	for {
+		select {
+		case data := <-oq.queue:
+			timeout := oq.initialTimeout
+			for try := 0; try < oq.maxRetries; try++ {
+				// we only care about last data
```

Comment:
Because previous data will be overwritten with latest data anyway, so there is no point in writing something if we already know that we will overwrite it in the next call (this is why it is called __*Overwrite*__ Queue 😄 )

Reply:
Got it. Thanks for the explanation!